beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java

```diff
@@ -312,19 +321,44 @@ public SafeFuture<Optional<BlockContainer>> createUnsignedBlock(
     if (isSyncActive()) {
       return NodeSyncingException.failedFuture();
     }
+    final Optional<BlockProductionPerformance> blockProductionPerformance =
+        blockProductionPerformanceTrackingEnabled
+            ? Optional.of(new BlockProductionPerformance(timeProvider, slot))
+            : Optional.empty();
     return forkChoiceTrigger
-        .prepareForBlockProduction(slot)
+        .prepareForBlockProduction(slot, blockProductionPerformance)
         .thenCompose(__ -> combinedChainDataClient.getStateAtSlotExact(slot))
+        .thenPeek(
+            maybeState -> {
+              if (blockProductionPerformance.isEmpty()) {
+                return;
+              }
+              maybeState.ifPresent(
+                  state ->
+                      blockProductionPerformance
+                          .get()
+                          .slotTime(secondsToMillis(spec.computeTimeAtSlot(state, slot))));
+              blockProductionPerformance.get().getStateAtSlot();
+            })
```

Comment:
If using noop pattern in tracker it may look something like this.  
```java
        .thenPeek(
            maybeState -> {
               blockProductionPerformance.slotTime(() -> {maybeState.map(state -> secondsToMillis(spec.computeTimeAtSlot(state, slot))}; 
               blockProductionPerformance.getStateAtSlot()
            })
```
And cleaner in other places. It would be good to not affect code readability much with tracking and metrics and better to put more logic in tracker than here

Reply:
not sure if it slides into metrics, it'd be neat if it does... worst case i think something like trackers with minimal deps could be nice...

Reply:
I agree on the NOOP pattern